### ISSUE TITLE
OpenID Connect authentication doesn't respect proxy settings

### ISSUE THREAD

**Issue by banderson-cbt** [Ylianst/MeshCentral#6616]
**Describe the bug**
OpenID Connect does not work through our HTTPS proxy. We are in a closed network where outbound connections are blocked, and HTTPS traffic must flow through a Web proxy.

For other things (npm, MeshCentral updates, timestamp server...) the proxy works fine.

In our environment, we had `agentTimeStampProxy` set to the proxy server, and the `http_proxy` and `https_proxy` environment variables set. I had tried setting `npmProxy` as well but that had no effect.

Upon starting the server, I get the following error message in the log, indicating that it's trying to reach the OIDC endpoint directly.

<details><summary>mesherrors.txt</summary>
-------- 12/18/2024, 5:12:53 PM ---- 1.1.35 --------

C:\meshcentral\node_modules\meshcentral\webserver.js:7498
                let error = new Error('OIDC: Discovery failed.', { cause: err });
                            ^

Error: OIDC: Discovery failed.
    at setupDomainAuthStrategy (C:\meshcentral\node_modules\meshcentral\webserver.js:7498:29)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async setupAllDomainAuthStrategies (C:\meshcentral\node_modules\meshcentral\webserver.js:6571:82) {
  [cause]: AggregateError [ETIMEDOUT]: 
      at internalConnectMultiple (node:net:1116:18)
      at afterConnectMultiple (node:net:1683:7) {
    code: 'ETIMEDOUT',
    [errors]: [
      Error: connect ETIMEDOUT 40.126.28.18:443
          at createConnectionError (node:net:1646:14)
          at Timeout.internalConnectMultipleTimeout (node:net:1705:38)
          at listOnTimeout (node:internal/timers:575:11)
          at process.processTimers (node:internal/timers:514:7) {
        errno: -4039,
        code: 'ETIMEDOUT',
        syscall: 'connect',
        address: '40.126.28.18',
        port: 443
      },
      Error: connect ETIMEDOUT 20.190.135.5:443
          at createConnectionError (node:net:1646:14)
          at Timeout.internalConnectMultipleTimeout (node:net:1705:38)
          at listOnTimeout (node:internal/timers:575:11)
          at process.processTimers (node:internal/timers:514:7) {
        errno: -4039,
        code: 'ETIMEDOUT',
        syscall: 'connect',
        address: '20.190.135.5',
        port: 443
      },
      Error: connect ETIMEDOUT 40.126.28.22:443
          at createConnectionError (node:net:1646:14)
          at Timeout.internalConnectMultipleTimeout (node:net:1705:38)
          at listOnTimeout (node:internal/timers:575:11)
          at process.processTimers (node:internal/timers:514:7) {
        errno: -4039,
        code: 'ETIMEDOUT',
        syscall: 'connect',
        address: '40.126.28.22',
        port: 443
      },
      Error: connect ETIMEDOUT 40.126.28.12:443
          at createConnectionError (node:net:1646:14)
          at Timeout.internalConnectMultipleTimeout (node:net:1705:38)
          at listOnTimeout (node:internal/timers:575:11)
          at process.processTimers (node:internal/timers:514:7) {
        errno: -4039,
        code: 'ETIMEDOUT',
        syscall: 'connect',
        address: '40.126.28.12',
        port: 443
      },
      Error: connect ETIMEDOUT 40.126.28.11:443
          at createConnectionError (node:net:1646:14)
          at Timeout.internalConnectMultipleTimeout (node:net:1705:38)
          at listOnTimeout (node:internal/timers:575:11)
          at process.processTimers (node:internal/timers:514:7) {
        errno: -4039,
        code: 'ETIMEDOUT',
        syscall: 'connect',
        address: '40.126.28.11',
        port: 443
      },
      Error: connect ETIMEDOUT 40.126.7.32:443
          at createConnectionError (node:net:1646:14)
          at Timeout.internalConnectMultipleTimeout (node:net:1705:38)
          at listOnTimeout (node:internal/timers:575:11)
          at process.processTimers (node:internal/timers:514:7) {
        errno: -4039,
        code: 'ETIMEDOUT',
        syscall: 'connect',
        address: '40.126.7.32',
        port: 443
      },
      Error: connect ETIMEDOUT 40.126.28.23:443
          at createConnectionError (node:net:1646:14)
          at Timeout.internalConnectMultipleTimeout (node:net:1705:38)
          at listOnTimeout (node:internal/timers:575:11)
          at process.processTimers (node:internal/timers:514:7) {
        errno: -4039,
        code: 'ETIMEDOUT',
        syscall: 'connect',
        address: '40.126.28.23',
        port: 443
      },
      Error: connect ECONNREFUSED 40.126.28.21:443
          at createConnectionError (node:net:1646:14)
          at afterConnectMultiple (node:net:1676:16) {
        errno: -4078,
        code: 'ECONNREFUSED',
        syscall: 'connect',
        address: '40.126.28.21',
        port: 443
      }
    ]
  }
}
</details> 

*Note:* We were able to resolve this using the third-party tool Proxifier, so we're not in a hurry to get this fixed. But I'm still reporting it here as an official solution would be ideal. 

**To Reproduce**
1. Operate in a closed network (outbound connections blocked by default)
2. Configure MeshCentral to use proxy server
3. Configure OIDC authentication strategy
4. Start server

**Expected behavior**
The server would start and OIDC authentication would be enabled

**Server Software (please complete the following information):**
 - OS: Windows Server 2019
 - Virtualization: Hyper-V
 - Network: Closed network w/ proxy server for outbound HTTP/S connections. Clients all on internal network and reachable directly
 - Version: 1.1.35
 - Node: 20.12.1

**Your config.json file**
```
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "cert": "meshcentral.company.com",
    "WANonly": true,
    "port": 443,
    "redirPort": 80,
    "exactPorts": true,
    "webrtc": false,
    "lockAgentDownload": true,
    "agentSignLock": true,
    "agentTimeStampServer": "http://timestamp.digicert.com",
    "_note": "We are using Proxifier for the time being, so proxy settings are commented out.",
    "_npmProxy": "http://proxy.company.com",
    "desktopMultiplex": true
  },
  "domains": {
    "": {
      "newAccounts": false,
      "amtManager": false,
      "amtScanner": false,
      "novnc": false,
      "mstsc": false,
      "ssh": false,
      "authStrategies": {
        "oidc": {
          "client": {
            "client_id": "[redacted]",
            "client_secret": "[redacted]"
          },
          "custom": {
            "preset": "azure",
            "tenant_id": "[redacted]"
          },
          "groups": {
            "recursive": true,
            "required": ["MeshCentral Users"],
            "siteadmin": ["MeshCentral Admins"],
            "revokeAdmin": true,
            "sync": {
              "filter": [
                "MeshCentral Admins",
                "MeshCentral - Information Technology",
                "MeshCentral - Electronic Banking",
                "MeshCentral - View Only"
              ]
            }
          },
          "newAccounts": true
        }
      },
      "newAccountsRights": [
        "nonewgroups",
        "nonewdevices",
        "nomeshcmd"
      ],
      "passwordRequirements": {
        "_comment__": "MFA will be enforced by the IdP",
        "allowaccountreset": false,
        "autofido2fa": false,
        "email2factor": false,
        "sms2factor": false,
        "push2factor": false,
        "otp2factor": false,
        "msg2factor": false,
        "backupcode2factor": false,
        "single2factorWarning": false,
        "force2factor": false,
        "loginTokens": false
      },
      "userConsentFlags": {
        "desktopnotify": true,
        "terminalnotify": true,
        "filenotify": true,
        "desktopprompt": true,
        "terminalprompt": true,
        "fileprompt": true,
        "desktopprivacybar": true
      },
      "autoRemoveInactiveDevices": 14
    }
  }
}
```

**Comment by si458** [comment]
so after quickly looking at the source code for `openid-client`
they dont actually use any 3rd party librarys to talk to the internet,
they use the normal `https.request` thats built into nodejs
https://github.com/panva/openid-client/blob/45c96f67ce0644bd829f61e82fba3dd8c051c89e/lib/helpers/request.js#L122-L133
https://nodejs.org/docs/latest-v22.x/api/https.html#httpsrequestoptions-callback
so from my understanding you should be able to use the HTTPS_PROXY environment variables?
are you sure you have set them up correctly with urls and usernames/password etc?

**Comment by si458** [comment]
ok just wanted to check! 👍 (sounded daft sorry) 

i did find an article that explains about building it into the https.request so might need investigating more
https://antoinevastel.com/nodejs/2022/02/26/nodejs-optimized-https-proxy-no-dependencies.html

i also found an old version of `openid-client` version v2.x had the option for proxies as it depended on an external library
https://github.com/panva/openid-client/issues/22
use this link from the issue instead 
https://github.com/panva/openid-client/tree/v2.x?tab=readme-ov-file#changing-http-request-defaults

but oversally we now use v5.x instead and these default options have now vanished so might of moved elsewhere

still looking into it for you tho!

**Comment by banderson-cbt** [comment]
Thanks, appreciate the help.

As mentioned we did work around it with a 3rd-party tool, so if it turns out to be a huge undertaking that it's not worth it, I won't be offended. :)

**Comment by si458** [comment]
ok so it appears we already have code in the `authenticode.js` which rewrites the `https.request` to use a proxy if you specify one
so in this case we just need to do the same for `openid-client` !
or at least work out HOW to pass the values to it so it will use the webproxy instead!

**Comment by banderson-cbt** [comment]
bump

**Comment by stephannn** [comment]
Hi @banderson-cbt ,
would you mind share some information about the third-party tool? Having the same issue unfortunately :(

**Comment by banderson-cbt** [comment]
The tool is called Proxifier. It costs money but it is cheap and a trial is available.

You'll need to install Proxifier in service mode. Create a profile using the GUI, then load it into the service. Exit the Proxifier GUI and start the service.

Below is what my Proxifier rules look like:

<img width="964" height="630" alt="Image" src="https://github.com/user-attachments/assets/7354313d-3a61-45bb-8ceb-5971d9d35c1f" />

**Comment by stephannn** [comment]
I see, thanks @banderson-cbt . I tried a proxy container but this seems to include iptables and a bit of too much problems in the future. We will try to put the login.microsoft.com and maybe more sites on the default route.
It is really an annoying issue. I checked all day and everything worked, so I could not imagine that the openid-client ignores the http_proxy variable.

**Comment by stephannn** [comment]
I think I could solve it that way:
1. Add the **proxy-agent** dependency (\opt\meshcentral\package.json) 

```
{
  "dependencies": {
    "connect-flash": "0.1.1",
    "meshcentral": "^1.1.50",
    "mongodb": "4.17.2",
    "nedb": "^1.8.0",
    "openid-client": "5.7.1",
    "otplib": "12.0.1",
    "passport": "0.7.0",
    "proxy-agent": "6.5.0"
  }
}
```

2. Install it:
`podman exec -it meshcentral npm install`

3. Modify the webserver.js
```
// Discover additional information if available, use endpoints from config if present
            let issuer
            try {
                parent.authLog('setupDomainAuthStrategy', `OIDC: Discovering Issuer Endpoints: ${strategy.issuer.issuer}`);
                //issuer = await strategy.obj.openidClient.Issuer.discover(strategy.issuer.issuer);
                const ProxyAgent = require('proxy-agent');

                const proxyUrl = process.env.HTTPS_PROXY || process.env.HTTP_PROXY;
                const agent = proxyUrl ? new ProxyAgent.ProxyAgent(proxyUrl) : undefined;
                
                issuer = await strategy.obj.openidClient.Issuer.discover(
                    strategy.issuer.issuer,
                    { agent }
                );
            }
```

Maybe someone, who understand meshcentral better can verify it

**Comment by si458** [comment]
ok so ive just done testing and i got `https-proxy-request` to work instead of `proxy-agent`
```
const { HttpsProxyAgent } = require('https-proxy-agent');
const proxyAgent = new HttpsProxyAgent('http://192.168.170.6:8080', { rejectUnauthorized: false });
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
strategy.obj.openidClient.custom.setHttpOptionsDefaults({ agent: proxyAgent });
...
...
issuer = await strategy.obj.openidClient.Issuer.discover(strategy.issuer.issuer);
```
BUT im now getting a warning in nodejs as expected
`Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.`

what proxy software do u use? and im assuming you use `https://my.auth.domain.com` for the issuer?
im currently using mitmproxy for my testing 
edit: ive also never used a forward proxy before, only every reverse proxies, so they are all new to me
(but i understand how they work etc)

EDIT: ok i got it to work without the reject_unauthorized, i had to set an environment variable for my CA of the mitmproxy,
then run the app `export NODE_EXTRA_CA_CERTS="/home/simon/meshcentral/meshcentral-data/mitmproxy-ca-cert.pem"`

**Comment by stephannn** [comment]
@si458 , yes, it  sounded like a cert issue.

I am not sure which proxy we are using to be honest. It is maintained by an external service provider, but I also think most proxies used in a corporate network are only HTTP and have no certificate. It leaves me to the question why, but on the other hand our clients do not accept proxies by DHCP for example.

**Comment by si458** [comment]
@stephannn ok so for urself how do u normally setup a proxy for an app? 
Just give it the url of the proxy?
U don't need to add like ssl certs or anything?

**Comment by stephannn** [comment]
@si458 , I talked to a colleague and he said http is more like a standard, but I am also not sure. Anyways, both should be covered. I modified your code a bit to get the code from the env variable:
package.json
```
{
  "dependencies": {
    ...
    "https-proxy-agent": "7.0.6"
  }
}
```

```

const { HttpsProxyAgent } = require('https-proxy-agent');
const proxyUrl = process.env.HTTPS_PROXY || process.env.HTTP_PROXY;
const proxyAgent = new HttpsProxyAgent(proxyUrl, { rejectUnauthorized: false });
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
strategy.obj.openidClient.custom.setHttpOptionsDefaults({ agent: proxyAgent });
```
and it works, I just receive a warning abouot TLS:
```
(node:14) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
(Use `node --trace-warnings ...` to show where the warning was created)
```
by the way, I set both variable HTTPS_PROXY and HTTP_PROXY in the docker container, but both are linked to our HTTP proxy

**Comment by si458** [comment]
@stephannn ok that's good news! 
What i might do is change the code so it checks if the http or https environment variables are set and if so then apply them. Sadly the isn't anything I can do about the warning as this is hard encoded in nodejs.
The only way around that as i explained was to not set the rejectunauthorised: false but instead specify the ca cert of my mitmproxy and then the warning never showed and it worked a treat!

Also weirdly enough meshcentral installs the https-proxy-agent if u set a proxy for the plugins!

So ill add another check that if again the environment variables are set for the proxies then it will load the module and away it goes!

**Comment by banderson-cbt** [comment]
> [@stephannn](https://github.com/stephannn) ok so for urself how do u normally setup a proxy for an app? Just give it the url of the proxy? U don't need to add like ssl certs or anything?

There's two ways to implement a HTTPS proxy, one is with SSL decryption (this requires the proxy server's cert to be installed on the device) or a CONNECT tunnel. A CONNECT tunnel relays the traffic, and the proxy only sees basic information (i.e. Host: header).

For a basic CONNECT tunnel one only needs to supply the proxy address.

**Comment by si458** [comment]
@banderson-cbt thank u for ur input!
I actually found a work around for the mitmproxy app
You had to use the ignore-hosts option and set it to * which simply tells it and ssl ignore and just use a tcp connect button still decrypt http!
So using this mode I don't need the ssl ca cert
But if u oversally need the ca cert then u add the extra environment variable!
Im going to try do a fix tomorrow morning for people to test, it will simply check if u have the http/https proxy environment variable set then it will install the https proxy agent module and tell oidc to use it :)

**Comment by stephannn** [comment]
@si458 , sorry to bother you again. I think I have a follow up error. After implementing your change, I can boot up MeshCentral (before, it was looping because it could not reach the Microsoft servers).
Now when I click on login I get following error:
```
Error: OIDC: GROUPS: Request error.
    at ClientRequest.<anonymous> (/opt/meshcentral/node_modules/meshcentral/webserver.js:8172:41)
    at ClientRequest.emit (node:events:519:28)
    at emitErrorEvent (node:_http_client:105:11)
    at TLSSocket.socketErrorListener (node:_http_client:518:5)
    at TLSSocket.emit (node:events:519:28)
    at emitErrorNT (node:internal/streams/destroy:170:8)
    at emitErrorCloseNT (node:internal/streams/destroy:129:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:90:21) {
  [cause]: AggregateError [ETIMEDOUT]: 
      at internalConnectMultiple (node:net:1134:18)
      at internalConnectMultiple (node:net:1210:5)
      at Timeout.internalConnectMultipleTimeout (node:net:1742:5)
      at listOnTimeout (node:internal/timers:590:11)
      at process.processTimers (node:internal/timers:523:7) {
    code: 'ETIMEDOUT',
    [errors]: [
      [Error], [Error], [Error],
      [Error], [Error], [Error],
      [Error], [Error], [Error],
      [Error], [Error], [Error],
      [Error], [Error], [Error],
      [Error]
    ]
  }
}
```

The code line is now incorrect, but it is the function: _async function getGroups_
Can you tell me how to dig into it? I found this env on the internet:
[--network-family-autoselection-attempt-timeout=500](url)
but no change

Edit:
ok, I did the same here

```
async function getGroups(preset, tokenset) {
                    let url = '';
                    const { HttpsProxyAgent } = require('https-proxy-agent');
                    if (preset == 'azure') { url = strategy.groups.recursive == true ? 'https://graph.microsoft.com/v1.0/me/transitiveMemberOf?$top=999' : 'https://graph.microsoft.com/v1.0/me/memberOf?$top=999'; }
                    if (preset == 'google') { url = strategy.custom.customer_id ? 'https://cloudidentity.googleapis.com/v1/groups?parent=customers/' + strategy.custom.customer_id : strategy.custom.identitysource ? 'https://cloudidentity.googleapis.com/v1/groups?parent=identitysources/' + strategy.custom.identitysource : null; }
                    return new Promise((resolve, reject) => {
                        const options = {
                            'headers': { authorization: 'Bearer ' + tokenset.access_token },
                            agent: process.env.HTTPS_PROXY ? new HttpsProxyAgent(process.env.HTTPS_PROXY) : undefined
                        }
```

I added
`agent: process.env.HTTPS_PROXY ? new HttpsProxyAgent(process.env.HTTPS_PROXY) : undefined`
in the options and I can login

**Comment by si458** [comment]
@stephannn beat me too it! i was just looking at it for you when i noticed your editted comment

i only SKIMMED the sourcecode originally and worked out how to get it to work
so chances are the is LOADS of other places that need to apply the https-proxy-agent too!

but thank you for spotting this one, i will sort it out next month ready for next months release for you 👍

**Comment by stephannn** [comment]
@si458 , thanks a lot, I really appreciate it.
We are doing a POC at work for MeshCentral so I am trying to dig out some stuff. Next point on my bucket list is the setting _manageAllDeviceGroups_ in combination with Azure groups ;) but for another issue I already saw ;)

**Comment by si458** [comment]
finally added support! https://github.com/Ylianst/MeshCentral/commit/1bd06ccded89bd48e2264d18117066bcd8f35bbe
thank you @stephannn for all the help and im sorry it took so long to add!
ive also added the proxy as an object in the meshcentral side so its ready for other places we have missed!
